Added 'is_landlocked' property

jav7zaid opened the pull request:
Issue [#47 ](https://github.com/bhatvikrant/world-countries-capitals/issues/47)

#Pull Request description

- Added new property is_landlocked in data.json file **source - wikipedia and world map **
- Added new method getCountriesByLandLock() to fetch the data in index.js
- Added test cases to cover the method in getCountriesSpec.js
- Fixed existing test cases.
- Updated readme.md for the new changes.
- Fixed a typo for getCountriesByFamousFor()

sthiepaan commented on README.md:
````diff
@@ -266,6 +266,38 @@ This section contains nt to _contribute_ to open source? Try solving our issues
+  This method returns an array of objects, each object containing **country**, **capital**, **currency**, **native_language**, **famous_for**, **phone_code**, **flag**, **drive_direction** and **is_landlocked** sorted by the _isLandlocked property which is boolean_ passed as the parameter.<br>
+  Response for getCountriesByLandLock(true)
+
+      	```javascript
+      	[
+      		{
+				country: 'afghanistan',
+				capital: 'kabul',
+				currency: 'afghani',
+				native_language: ['dari persian', 'pashto'],
+				famous_for: 'rugs, taliban',
+				phone_code: '+93',
+				flag: 'https://flagpedia.net/data/flags/h80/af.png',
+				drive_direction: 'right',
+				is_landlocked: true
+  			},
+      		{
+      			country: 'andorra',
+      			capital: 'andorra la vella',
+      			currency: 'euro',
+      			native_language: ['catalan'],
+      			famous_for: 'tourism and tobacco',
+				phone_code: '+244',
+				flag: 'https://flagpedia.net/data/flags/h80/ad.png',
+				drive_direction: 'right',
+    			is_landlocked: true
+      		}
+      	];
````
The indentation looks not so well. Could we make it better? 😉

jav7zaid replied:
Thanks, will do.

sthiepaan commented on index.js:
```diff
@@ -125,6 +125,20 @@ const getCountriesByFamousFor = (famousThing) => {
+/** 
+ * Returns an array of objects containing all countries, each containing `country`, `capital`,
+ * `currency`, `native_language`, `famous_for`, `phone_code`, `flag` and `drive_direction` filtered by `is_landlocked`
+ * @param {'true' | 'false'} isLandLocked  Country is said to be landlocked when it is surrounded on all sides by one or more countries
```
The way you defined that param would mean that user needs to provide one of two `String` but actually it has to be `Boolean`

jav7zaid replied:
Thanks ! will make the change.

sthiepaan commented on index.js:
```diff
@@ -125,6 +125,20 @@ const getCountriesByFamousFor = (famousThing) => {
+ * @returns {Array} An array of country objects
+*/
+const getCountriesByLandLock = (isLandLocked) => {
+  const isBoolean = typeof isLandLocked === 'boolean'
+  const filterBy = isBoolean && isLandLocked
```
I am not sure if it is so necessary to have that kind of checks. I think that proper documentation and `@param` Definition should be enough for the _API consumers_.

jav7zaid replied:
Just making sure that user only enters boolean, but i get your point the definition should be enough.
But in order to `through an error` i need to check type of the value.

jav7zaid replied:
If the user enters any value other than boolean, i will throw an error suggesting proper value.

sthiepaan commented on index.js:
```diff
@@ -125,6 +125,20 @@ const getCountriesByFamousFor = (famousThing) => {
+const getCountriesByLandLock = (isLandLocked) => {
+  const isBoolean = typeof isLandLocked === 'boolean'
+  const filterBy = isBoolean && isLandLocked
+  return isBoolean && data.filter( 
+    (country) => country.is_landlocked === filterBy
+  );
```
```suggestion
  return data.filter(country => country.is_landlocked === isLandLocked);
```

sthiepaan commented on package-lock.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "world-countries-capitals",
-	"version": "3.1.0",
+	"version": "3.2.1",
```
Why is this change within that _Pull Request_? Can we revert that from here?

@bhatvikrant is there any issue with package versioning over _NPM_? Was last release happened properly?

jav7zaid replied:
I think it got staged somehow and pushed, i will rather add it in gitignore so that it doesn't comes up in my next PR.

bhatvikrant replied:
No there is no issue with versioning, @jav7zaid lets not update the package versions in PRs. I ll update versions while publishing to NPM

sthiepaan commented on spec/getCountriesSpec.js:
```diff
@@ -25,6 +25,7 @@ describe("The index", () => {
         phone_code: "+93",
         flag: "https://flagpedia.net/data/flags/h80/af.png",
         drive_direction: "right",
+        is_landlocked: true
```
Please add comma at the end of line so other does not have to take care of line created by you 😉
```suggestion
        is_landlocked: true,
```
Same for lines (28, 44, 66, 161, 175)

jav7zaid replied:
Thanks! will take care

sthiepaan commented on spec/getCountriesSpec.js:
```diff
@@ -135,4 +138,46 @@ describe("The index", () => {
+
+  describe("The getAllCountriesByLandlock", () => {
+    it("returns all the landlocked countries", () => {
+      expect(countryApi.getCountriesByLandLock(true).length).toEqual(45);
```
Can we move the `countryApi.getCountriesByLandLock(true)` (and the one with `false`) so it can be reused in other specs? (same for line: 144, 148, 152, 166)

jav7zaid replied:
Agreed will do.

bhatvikrant replied:
Yeaa, that will be neat

sthiepaan commented on spec/getCountriesSpec.js:
```diff
@@ -135,4 +138,46 @@ describe("The index", () => {
+    });
+
+    it("should return undefined if passed any value other than boolean", () => {
+      expect(countryApi.getCountriesByLandLock(1).length).toEqual(undefined);
```
From my point of view, there are two ways, how _API_ should behave in that case:
1. Throw an error that suggest what is the proper value
2. Return empty array

From a user perspective, 1. seems to be more handy, since it tells me what I did wrong and help me without checking the documentation.

jav7zaid replied:
Yes i agree throwing an error would be suffice, will make the changes.

sthiepaan reviewed (changes requested):
Looks promising! I added few comments/suggestions to improve your code 😉
Thank you @jav7zaid!

sthiepaan commented:
@jav7zaid you don't need to answer to every single comment - it spams my e-mail so hard 😅... If you decide that you will apply changes, you can just do them and after you finish/commit mark a comment as resolved 😉

jav7zaid commented:
@bhatvikrant and @sthiepaan please check the PR ,i have updated it with review comments.
I have removed the package-lock.json file.
Let me know if looks good to be merged.

sthiepaan commented on index.js:
```diff
@@ -128,15 +128,11 @@ const getCountriesByFamousFor = (famousThing) => {
  * Returns an array of objects containing all countries, each containing `country`, `capital`,
  * `currency`, `native_language`, `famous_for`, `phone_code`, `flag` and `drive_direction` filtered by `is_landlocked`
- * @param {'true' | 'false'} isLandLocked  Country is said to be landlocked when it is surrounded on all sides by one or more countries
+ * @param {true | false} isLandLocked  Country is said to be landlocked when it is surrounded on all sides by one or more countries
```
```suggestion
 * @param {Boolean} isLandLocked  Country that is surrounded by one or more countries
```

sthiepaan commented:
@jav7zaid after you finish applying and pushing changes - instead of writing a comment - you can click on little circle icon in _Reviewers_ section (located at top right of the _Pull Request_) near reviewer nickname 😉

sthiepaan reviewed (approved):
LGTM

bhatvikrant reviewed (changes requested):
@jav7zaid please resolve conflicts and then its good to go!

sthiepaan reviewed (changes requested):
I would like to merge this _Pull Request_ but unfortunately we struggle with some merge conflicts. Conflicts are in:
- `README.md`
- `data/data.json`
- `index.js`
- `spec/getCountriesSpec.js`

@jav7zaid could you resolve them? 😉

jav7zaid commented:
> I would like to merge this _Pull Request_ but unfortunately we struggle with some merge conflicts. Conflicts are in:
> 
> * `README.md`
> * `data/data.json`
> * `index.js`
> * `spec/getCountriesSpec.js`
> 
> @jav7zaid could you resolve them? 😉

done @sthiepaan

sthiepaan reviewed (approved):
LGTM

jav7zaid commented:
> > I would like to merge this _Pull Request_ but unfortunately we struggle with some merge conflicts. Conflicts are in:
> > 
> > * `README.md`
> > * `data/data.json`
> > * `index.js`
> > * `spec/getCountriesSpec.js`
> > 
> > @jav7zaid could you resolve them? 😉
> 
> done @sthiepaan

i think my changes disturbed the existing test cases fixing them.

sthiepaan commented:
@jav7zaid actually yes, you broke it! 😅... Always run `npm test` before making a commit.
Please make a new _Pull Request_ with fix for that _ASAP_.
